### DISCUSSION TITLE
HELP! Can not load CTB file(about 4GB 20000layers) (ChituBox 1.7.0 and 1.93) into UVTools3.8.3

### DISCUSSION THREAD

**Discussion by vadimav1** [sn4k3/UVtools#608] [Q&A]
HELP! Can not load CTB file  into UVTools!
Small *.CTB file examples are loading with success. 
But, real large print file load process leads to ERRORS!
It contains about 4GB and  about 20000 layers.
I try to generate it from ChituBox 1.7.0 and 1.93. (crypted and non-crypted)
At about 70% of progress bar indicator of loading i see the errors. 
I tried to use 3.5.6 and 3.8.3 UVTools.
I checked it by ChituBox. It reloaded it from CTB to 3D-look back successfully.
I see no overloading of PC-AT  Win10-64 RAM (64G) or HDD. 

See the Errors:

![изображение](https://user-images.githubusercontent.com/118778617/203178437-e2fdcc08-ff78-4588-b4dd-9a2cb7767df3.png)

![изображение](https://user-images.githubusercontent.com/118778617/203177298-6d94a26a-ef8c-4351-924e-9ea8421b77b3.png)

**Comment by sn4k3** [comment]
4GB and about 20000 layers 😱

Thats way too much! What is your layer height?
Not impossible to load but that require a large amount of RAM on your machine, UVtools caches images on RAM, if you have not enough RAM such problems happens.

**Reply by vadimav1** [reply]
I have 64G of RAM. UVTools used about 10GB. A lot of free RAM there. But error occurs! 
I use 3d voxel  mathematics and use 1mkm layer thickness. 
My aim - to extract PNG files only from CTB, using UVTools, because CTB file is not a simply ZIP.

**Reply by sn4k3** [reply]
1mkm is that 0.01mm?

If you want futher check please upload the file somewhere and send to me

**Reply by sn4k3** [reply]
> My aim - to extract PNG files only from CTB, using UVTools, because CTB file is not a simply ZIP.

Chitubox can export to Zip if you use a custom defined printer, that way you can bypass UVtools step

**Reply by vadimav1** [reply]
1mkm = 0.001mm a thoused microns in 1milimeter.

**Reply by vadimav1** [reply]
> Chitubox can export to Zip if you use ... 

Yes, but ChituBox save ZIP file exlremely long time (about a half day)
On the contrary,  CTB file is saving during some minutes.
Therefore i try to use CTB format for ChituBox saving in aim to use it in UVTools.

**Reply by vadimav1** [reply]
Then UVTools is used to get PNGs

**Reply by sn4k3** [reply]
Have you considered Lychee? It have fast export to zip.
Otherwise send me the file so I can track down the problem and replicate the issue

**Reply by vadimav1** [reply]
My aim - get all  PNGs  from Chitu Slicer.  
After this i launch my own mathematic to get 3D Voxels  and resin normalization.
Then, i plan  to obtain standart print  file with layer thickness = 20micron. (20 times thicker) and print it.

**Reply by vadimav1** [reply]
Even begin with 15-20 micron thickness  cappilary forces will allow to  press resin  for real layer.

**Reply by vadimav1** [reply]
No. i m not so advanced in 3D software. i used perfactory before. but it generate  a lot of gel fishes in layer. 
QUESTION1 - does Lychee allow to adjust layer thickness up to 1mkm (0.001mm) ? 
QUESTION2- how i to thanslate a test file to You?

**Reply by vadimav1** [reply]
Earlier I used 3D photopolymer 2K printer with open marlin G-Code architecture and opened print file format (Zip with PNGs and text file with G-Codes)
But now i need more larger field of printing. 
Only Chitu printer is available, but it have closed (crypted) format. So, I have a big headpain how to adjust it!

**Reply by sn4k3** [reply]
> Even begin with 15-20 micron thickness cappilary forces will allow to press resin for real layer.

I think less than 30 μm there are no visual benefits for us.

QUESTION1 - Looks like yes:
![LycheeSlicer_2022-11-22_00-01-20](https://user-images.githubusercontent.com/113281/203182253-68d488cb-5ee8-4639-b22f-5d4bb14fc84c.png)

QUESTION2- Try with: https://easyupload.io/

> Only Chitu printer is available, but it have closed (crypted) format. So, I have a big headpain how to adjust it!

With chitubox you can also try Phrozen Transformer, i think that one use a pure Zip format

**Reply by vadimav1** [reply]
I try to save to  Phrozen format  from ChituBox. It saved fastly, but UVTools destroy PNGs It's look like a noise. 
QUESTION3 - does Lychee allow to save CTB files? (in aim to compose for my new 8K printer) ? 
QUESTION4 - how to get Lychee slicer program?

**Reply by vadimav1** [reply]
For example, i simply consreuct by G codes and few images a simply  exposure tester for my old printer, But how to create this for chitu printer?  See layer example. https://www.youtube.com/shorts/Num7hfL1kWE
It prints 30 squares with step  for each 0.5 sec.   Wholle exposure finder for a short time.

**Reply by vadimav1** [reply]
CDBDLP format also leads to errors in UV tools.  Probably inside size restrictions or layer counter variable overflow.
![изображение](https://user-images.githubusercontent.com/118778617/203185102-bf756a54-4047-4668-a448-9734d43a5025.png)

**Reply by vadimav1** [reply]
You said: I think less than 30 μm there are no visual benefits for us.

But it requires to use tilt for detail placing , in aim to avoid horisontal planes. 
But tilt leads to tilt-texture at details. This is not allowable for machinery details.
Texture along axle of printer and in axles of machinery detail is applicable. 
But at horisontaly -like planes we will have visible Z-step projections  which will have visible step. 
To avoid this (and XY such effects ) i developed 3D Voxel with resin normalisation approach. 
Also it allow to sufficiently diminish time of printing (no tilts for flat details)

I think, now You start understanding me.

**Reply by vadimav1** [reply]
I load test CTB file  into https://easyupload.io/znharq
PWD=1234
at 7 days

**Reply by sn4k3** [reply]
> I try to save to Phrozen format from ChituBox. It saved fastly, but UVTools destroy PNGs It's look like a noise.

Then you don't need UVtools, you just open the file as Zip and extract images.

QUESTION3 - Yes
QUESTION4 - https://mango3d.io/downloads

> For example, i simply consreuct by G codes and few images a simply exposure tester for my old printer, But how to create this for chitu printer? See layer example. https://www.youtube.com/shorts/Num7hfL1kWE
It prints 30 squares with step for each 0.5 sec. Wholle exposure finder for a short time.

You can't with actual machines, none is capable of that, you always require to change the whole image and a lift in between images. Only gcode printers can ignore lifts but still require image change.

> CDBDLP format also leads to errors in UV tools. 

Forget cbdlp, is awful format, please dont use it.

> Probably inside size restrictions or layer counter variable overflow.

Size restriction is: 2,147,483,647

> I load test CTB file

Downloading, will report latter

**Reply by sn4k3** [reply]
Here what's happening:

```
layer[21610]: TableSize: 88, PositionZ: 21.611, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4291829561, Unknown2: 0, DataLength: 247520, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247520
layer[21611]: TableSize: 88, PositionZ: 21.612, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4292077169, Unknown2: 0, DataLength: 247525, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247525
layer[21612]: TableSize: 88, PositionZ: 21.613, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4292324782, Unknown2: 0, DataLength: 247402, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247402
layer[21613]: TableSize: 88, PositionZ: 21.614, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4292572272, Unknown2: 0, DataLength: 247339, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247339
layer[21614]: TableSize: 88, PositionZ: 21.615, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4292819699, Unknown2: 0, DataLength: 247927, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247927
layer[21615]: TableSize: 88, PositionZ: 21.616, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4293067714, Unknown2: 0, DataLength: 247781, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247781
layer[21616]: TableSize: 88, PositionZ: 21.617, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4293315583, Unknown2: 0, DataLength: 247903, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247903
layer[21617]: TableSize: 88, PositionZ: 21.618, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4293563574, Unknown2: 0, DataLength: 247967, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247967
layer[21618]: TableSize: 88, PositionZ: 21.619, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4293811629, Unknown2: 0, DataLength: 247895, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247895
layer[21619]: TableSize: 88, PositionZ: 21.62, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4294059612, Unknown2: 0, DataLength: 248043, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 248043
layer[21620]: TableSize: 88, PositionZ: 21.621, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4294307743, Unknown2: 0, DataLength: 247972, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247972
layer[21621]: TableSize: 88, PositionZ: 21.622, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4294555803, Unknown2: 0, DataLength: 247845, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247845
layer[21622]: TableSize: 88, PositionZ: 21.623, ExposureTime: 2, LightOffDelay: 0, LayerDefOffset: 4294803736, Unknown2: 0, DataLength: 247986, Unknown3: 0, EncryptedDataOffset: 0, EncryptedDataLength: 0, LiftHeight: 7, LiftSpeed: 30, LiftHeight2: 4, LiftSpeed2: 150, RetractSpeed: 150, RetractHeight2: 1, RetractSpeed2: 30, RestTimeBeforeLift: 0, RestTimeAfterLift: 0, RestTimeAfterRetract: 1, LightPWM: 255, Unknown6: 0, RLEData: 247986
layer[21623]: TableSize: 2097152000, PositionZ: 4.902234E-39, ExposureTime: 5.6294995E+14, LightOffDelay: 0, LayerDefOffset: 3120562176, Unknown2: 3499121, DataLength: 1476395008, Unknown3: 0, EncryptedDataOffset: 3674210304, EncryptedDataLength: 3499891, LiftHeight: 5.6294995E+14, LiftSpeed: 0, LiftHeight2: 0, LiftSpeed2: 4.905472E-39, RetractSpeed: 5.6294995E+14, RetractHeight2: 0, RetractSpeed2: 6.938894E-18, RestTimeBeforeLift: 4.906551E-39, RestTimeAfterLift: 5.6294995E+14, RestTimeAfterRetract: 0, LightPWM: -1.661535E+35, Unknown6: 3502201, RLEData: 1476395008
layer[21624]: TableSize: 0, PositionZ: 2.729534E+31, ExposureTime: 3.21E-43, LightOffDelay: 3.1568E-41, LayerDefOffset: 0, Unknown2: 2004286976, DataLength: 229, Unknown3: 22528, EncryptedDataOffset: 0, EncryptedDataLength: 2067847936, LiftHeight: 3.21E-43, LiftSpeed: 3.1568E-41, LiftHeight2: 0, LiftSpeed2: 1.8366192E+38, RetractSpeed: 3.21E-43, RetractHeight2: 3.1568E-41, RetractSpeed2: 0, RestTimeBeforeLift: -3.1167245E-37, RestTimeAfterLift: 3.21E-43, RestTimeAfterRetract: 3.1568E-41, LightPWM: 0, Unknown6: 2258540544, RLEData: 229
layer[21625]: TableSize: 954390147, PositionZ: 3.4499235E-12, ExposureTime: -4.2686004E+22, LightOffDelay: 16509.496, LayerDefOffset: 1376152231, Unknown2: 2610311279, DataLength: 1060385880, Unknown3: 748971521, EncryptedDataOffset: 3828088522, EncryptedDataLength: 1103731891, LiftHeight: 5.8460717E+10, LiftSpeed: -9.257014E-23, LiftHeight2: 0.20270891, LiftSpeed2: 4.3116327E-10, RetractSpeed: -3.0629562E+26, RetractHeight2: 7.6013527, RetractSpeed2: 7.900012E+10, RestTimeBeforeLift: -2.6143618E-26, RestTimeAfterLift: 0.26355463, RestTimeAfterRetract: 3.4257413E-11, LightPWM: -4.369731E+26, Unknown6: 1126490141, RLEData: 1060385880
layer[21626]: TableSize: 1792744278, PositionZ: -3.448088E+35, ExposureTime: -1.7427792E-19, LightOffDelay: -729.59247, LayerDefOffset: 3607080315, Unknown2: 2393358073, DataLength: 210770537, Unknown3: 1482281743, EncryptedDataOffset: 1107476639, EncryptedDataLength: 583754781, LiftHeight: 2.931667E+34, LiftSpeed: -1.22209436E+27, LiftHeight2: -3.7540304E-11, LiftSpeed2: -7.117422E-06, RetractSpeed: -3.1843597E+22, RetractHeight2: -1.1791991E-38, RetractSpeed2: 4.1705385E-23, RestTimeBeforeLift: 2911765, RestTimeAfterLift: 3.3568051E+10, RestTimeAfterRetract: 1.653984E-26, LightPWM: -4.45328E-35, Unknown6: 3730139023, RLEData: 210770537
layer[21627]: TableSize: 1061608668, PositionZ: -1.4417278E-10, ExposureTime: 264562830, LightOffDelay: -116.10813, LayerDefOffset: 1387462518, Unknown2: 4038240502, DataLength: 1625854243, Unknown3: 3866909195, EncryptedDataOffset: 1981678472, EncryptedDataLength: 340184156, LiftHeight: -2.5003342E-36, LiftSpeed: 6.9570384E-33, LiftHeight2: -6.135046E-14, LiftSpeed2: 0.000417583, RetractSpeed: -1.5066088, RetractHeight2: 2.9923342E-10, RetractSpeed2: -516501120, RestTimeBeforeLift: 1.0727529E+18, RestTimeAfterLift: -7.429476E+11, RestTimeAfterRetract: 1.462875E+31, LightPWM: -2.6105365E+20, Unknown6: 1627543135, RLEData: 1625854243
layer[21628]: TableSize: 2405702851, PositionZ: -1.194224E+13, ExposureTime: -2412.6226, LightOffDelay: -9.449281E-20, LayerDefOffset: 4220040932, Unknown2: 1768994420, DataLength: 825844506, Unknown3: 870500506, EncryptedDataOffset: 1739311112, EncryptedDataLength: 4254342568, LiftHeight: -2.9342808E-21, LiftSpeed: -34230.74, LiftHeight2: -1.7728208E+12, LiftSpeed2: -2.8997887E-28, RetractSpeed: 2.7060473E-33, RetractHeight2: 4.7600955E+16, RetractSpeed2: 0.5730658, RestTimeBeforeLift: 3.7937523E-16, RestTimeAfterLift: 3.330891E+32, RestTimeAfterRetract: -6.33803E+28, LightPWM: -6.459222E-13, Unknown6: 3120439066, RLEData: 825844506
Exception thrown: 'System.OverflowException' in UVtools.Core.dll
```

 ~~The file looks like corrupt, after layer **21622**, it start to return random values which make UVtools to read on bad memory and overflow. This looks like Chitubox problem and a limitation within thier software. Please try slice CTB on Lychee and report back.~~
PrusaSlicer can also slice zip files: [Guide](https://github.com/sn4k3/UVtools/wiki/Setup-PrusaSlicer)

**Reply by sn4k3** [reply]
I was able to identify the problem. It work correctly now:

![image](https://user-images.githubusercontent.com/113281/203214029-b2a09e88-099c-4233-8cfd-eaa229a9aa3a.png)

Will try to issue a patch.
It is because they use page files inside format because printers can't handle large numbers.

**Reply by sn4k3** [reply]
(If you have them) please send me same file in CTB (not encrypted format), CBDDLP and other formats you have tried for I implement page files in them too. Or better, if you can send me the STL who generate this slices so I can generate the files myself for all formats.

**Reply by vadimav1** [reply]
Hello again. 
EasyUpLoad does not allow to load more than 10G files .
CDBDLP, FDG , PHZ   files bigger than 10G

**Reply by vadimav1** [reply]
I think, You can get any STL with height about 20mm and slice it with 1mkm spacing. I have old PC, so, slice process is sufficiently long. FDG save process took me more than 24h and I canceled it, because common power feed was disrupted somewhere outside for a long period and UPC battery discharged.

**Reply by vadimav1** [reply]
> Will try to issue a patch. 
Where i can get a patch? How to get it?

**Reply by sn4k3** [reply]
> I think, You can get any STL with height about 20mm and slice it with 1mkm spacing.

Ok, i will try it.

> Where i can get a patch? How to get it?

When I release next version, in two days maybe

**Reply by vadimav1** [reply]
Hello again! 
How are You?
How about  solution of this problem?

**Reply by sn4k3** [reply]
It's solved on latest version, please update

**Reply by vadimav1** [reply]
Ok. Many Thanks!!!  New UVTools opened such big files and saved PNGs. Continue testing...

As to Exposure finder,  One can propose for each Z layer to imagine all 30 squares for 0.5sec, then remaining 29 squares for 0.5sec...  So, this would  work faster  in parallel in case of all 30 objects. 
What do You think about it? 

Another problem - how to shut off forced recalibration which occurs before printing  and deathly sewn into firmware bosoms. May be You allready enountered with it? : 

Detailed description: 

1) I use extra-resolution resins with hardening depth about 20-30 micron only!
 
2) To use and print with this resins, I use Z-axle special support with spring, which, being unfixed with table, press table down with correct force, when resin is placed in quvette and support is mooving  down into calibration marker and stop. Calibration position is done at this moment.
 
3) If i rigidly fix object table to z-axle support  by knob turn  and start print cycle immediately  - i will achieve no first layer attached to table, because resin layer is still too thick.
 
4) About  10 minutes is required to squeeze out resin, forming thight contact between table and bottom of сuvette. Thickness of resin layer is diminishing with time from 0.5mm to 20mkm.
(this velocities is not supported by forced calibration menu setup, and this is different for different resins, temperature...)
  
5) If i will use standart table where table is rigidly fixed on support constantly - i will  crack LCD screen or Z axle motor will loose steps, because i use extreme rigid Z-Axle and LCD on thick quarz.
 
6) Time to time, I measure the distance between cuvette bottom and table (remained resin layer thickness), and when this value becomes less than 20-10 microns, I fixed table to support by special knob, and want to start printing.
 
If recalibration will occur at this moment, i will crack LCD, or motor will loose steps, and, even will be ok wih LCD or motor - i do not achieve first layer attached to table.
 
7) I use this manual precise procedure because i use extra-resolution resins with hardening depth about 20-30micron only!
